popgen/nkpCacheIntegral.go

```diff
@@ -0,0 +1,34 @@
+package popgen
+
+import (
+	"github.com/vertgenlab/gonomics/numbers"
+	"math"
+)
+
+func LogIntegrateStationarityCache(alpha float64, n int, k int, stepSize int, alleleFrequencyCache []float64, nkpCache [][][]float64) float64 {
+	var bins int = len(nkpCache[n][k])-1
+	var deltaX float64 = (RightBound - LeftBound) / float64(bins)
+	var logDeltaX float64 = math.Log(deltaX)
+	var currLeftIndex int = 0
+	var currRightIndex int = stepSize
+	var answer, rightEval float64
+
+	//first tiem, sets answer as the area of the first rectangle
+	var nextLeftEval float64 = math.Log(AFSStationarity(alleleFrequencyCache[currRightIndex], nkpCache[n][k][currRightIndex]))
+	answer = f(alpha, currLeftIndex, alleleFrequencyCache, nkpCache[n][k][currLeftIndex])
+	firstLeft := math.Log(AFSStationarity(alleleFrequencyCache[currLeftIndex], nkpCache[n][k][currLeftIndex]))
+	answer = numbers.MultiplyLog(numbers.MidpointLog(firstLeft, nextLeftEval), logDeltaX)
+
+	for i := stepSize; i < bins; i += stepSize {
+		currLeftIndex += stepSize
+		currRightIndex += stepSize
+		rightEval = f(alpha, currRightIndex, alleleFrequencyCache, nkpCache[n][k][currRightIndex])
+		answer = numbers.AddLog(answer, numbers.MultiplyLog(numbers.MidpointLog(nextLeftEval, rightEval), logDeltaX))
+		nextLeftEval = rightEval
+	}
+	return answer
```

Comment:
The way you calculate answer currently is the most readable, but I think we can speed it up with some math tricks.  One thing to note is that adding in log-space is expensive because you need to do both a math.Exp and a math.Log to get the answer.
If we have 5 bins, then we will have step ~= 0.2 going between near-0 and near-1.  Right now we do:
(f(near-0)+f(0.2))/2 * 0.2 + (f(0.2)+f(0.4))/2 * 0.2 + (f(0.4)+f(0.6))/2 * 0.2 + (f(0.6)+f(0.8))/2 * 0.2 + (f(0.8)+f(near-1))/2 * 0.2
additions is: 2 * bins -1
we could do:
0.2 * (f(near-0)/2 + f(0.2) + f(0.4) + f(0.6) + f(0.8) + f(near-1)/2)
this comes out the same, but almost halves our additions

Reply:
So instead of keeping an sum of areas I can keep a sum of heights and multiply by the width at the end? I'll change variable names accordingly so we can get the speed up and keep some readabilitty